parser/src/cfg/chord.rs

```diff
@@ -10,105 +15,328 @@ pub(crate) fn parse_defchordv2(
     s: &ParserState,
 ) -> Result<ChordsForKeys<'static, KanataCustom>> {
     let mut chunks = exprs[1..].chunks_exact(5);
-    let mut all_chords = FxHashSet::default();
     let mut chords_container = ChordsForKeys::<'static, KanataCustom> {
         mapping: FxHashMap::default(),
     };
-    for chunk in chunks.by_ref() {
-        let keys = &chunk[0];
-
-        let mut participants = keys
-            .list(s.vars())
-            .map(|l| {
-                l.iter()
-                    .try_fold(vec![], |mut keys, key| -> Result<Vec<u16>> {
-                        let k = key.atom(s.vars()).and_then(str_to_oscode).ok_or_else(|| {
-                            anyhow_expr!(
-                                key,
-                                "The first chord item must be a list of keys.\nInvalid key name."
-                            )
-                        })?;
-                        keys.push(k.into());
-                        Ok(keys)
-                    })
-            })
-            .ok_or_else(|| anyhow_expr!(keys, "The first chord item must be a list of keys."))??;
-        if participants.len() < 2 {
-            bail_expr!(keys, "The minimum number of participating chord keys is 2");
-        }
-        participants.sort();
-        if !all_chords.insert(participants.clone()) {
-            bail_expr!(
-                keys,
+    let all_chords = chunks
+        .by_ref()
+        .flat_map(|chunk| match chunk[0] {
+            // Match a line like
+            // (include filename.txt) () 100 all-released (layer1 layer2)
+            SExpr::List(Spanned {
+                t: ref exprs,
+                span: _,
+            }) if matches!(exprs.first(), Some(SExpr::Atom(a)) if a.t == "include") => {
+                let file_name = exprs[1].atom(s.vars()).unwrap();
+                let chord_translation = ChordTranslation::create(
+                    file_name,
+                    &chunk[2],
+                    &chunk[3],
+                    &chunk[4],
+                    &s.layers[0][0],
+                );
+                let chord_definitions = parse_chord_file(file_name).unwrap();
+                let processed = chord_definitions.iter().map(|chord_def| {
+                    let chunk = chord_translation.translate_chord(chord_def);
+                    parse_single_chord(&chunk, s)
+                });
+                Ok::<_, ParseError>(processed.collect_vec())
+            }
+            SExpr::List(_) => Ok(vec![parse_single_chord(chunk, s)]),
+            _ => Ok(vec![]),
```

Comment:
Seems like this is silently ignoring config errors where the first element is an atom? Seems like undesirable behaviour; I think this should be combined with the line above, i.e.

```rust
            _ => Ok(vec![parse_single_chord(chunk, s)]),
```

Reply:
Done